tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,767 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package chunks
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/binary"
+	"hash"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/tsdb/chunkenc"
+	tsdb_errors "github.com/prometheus/prometheus/tsdb/errors"
+	"github.com/prometheus/prometheus/tsdb/fileutil"
+)
+
+// Head chunk file header fields constants.
+const (
+	// MagicHeadChunks is 4 bytes at the beginning of a head chunk file.
+	MagicHeadChunks = 0x0130BC91
+
+	headChunksFormatV1 = 1
+	writeBufferSize    = 4 * 1024 * 1024 // 4 MiB.
+)
+
+var (
+	// DefaultHeadChunkFileMaxTimeRange is the default head chunk file time range.
+	// Assuming a general scrape interval of 15s, a chunk with 120 samples would
+	// be cut every 30m, so anything <30m will cause lots of empty files. And keeping
+	// it exactly 30m also has a chance of having empty files as its near that border.
+	// Hence keeping it a little more than 30m, i.e. 40m.
+	DefaultHeadChunkFileMaxTimeRange = 40 * int64(time.Minute/time.Millisecond)
+
+	// ErrChunkDiskMapperClosed returned by any method indicates
+	// that the ChunkDiskMapper was closed.
+	ErrChunkDiskMapperClosed = errors.New("ChunkDiskMapper closed")
+)
+
+// corruptionErr is an error that's returned when corruption is encountered.
+type corruptionErr struct {
+	Dir       string
+	FileIndex int
+	Err       error
+}
+
+func (e *corruptionErr) Error() string {
+	return errors.Wrapf(e.Err, "corruption in head chunk file %s", segmentFile(e.Dir, e.FileIndex)).Error()
+}
+
+// ChunkDiskMapper is for writing the Head block chunks to the disk
+// and access chunks via mmapped file.
+type ChunkDiskMapper struct {
+	/// Writer.
+	dir *os.File
+
+	curFile         *os.File // File being written to.
+	curFileSequence int      // Index of current open file being appended to.
+	curFileMint     int64    // Current file's mint. Used to check for a chunk crossing the max file time range.
+	curFileMaxt     int64    // Maxt of all chunks written to current file. This is written to the file when the file is closed.
+	curFileNumBytes int64    // Bytes written in current open file.
+	maxFileMs       int64    // Max time range (curFileMaxt-curFileMint) for a file.
+
+	writeBuf     [MaxHeadChunkMetaSize]byte
+	wbuf         *bufio.Writer
```

Comment:
> if they have their own struct with methods

Went through this and I think it might not simplify it much. `curFile` and it's info (`curFileXXX` variables) is also used with the reader variables (like when cutting a new file). Will skip this grouping for now, but will change that variable name as suggested.